### ISSUE TITLE
Incorrect error when using generic, class and conditional typing

### ISSUE THREAD

**Issue by pjannesen** [microsoft/TypeScript#43237]
# Bug Report

Incorrect error is reported when using generic, class and conditional typing

### 🕗 Version & Regression Information

The incorrect error is introduced in version 3.9 and exists in nightly

### ⏯ Playground Link

[playground](https://www.typescriptlang.org/play?noLib=true&ts=4.2.3#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgSQDEFgAbAEwBFh0AoAb1rmbgG0lMBbYALjgGcYUZAHMAun0xIAngG5aAX1q1QkWHGylM-fnABKwbNHIAeACoHscUKiTldREhWroAfAyYswwgG6ZUcJgAFACUPBaGAD5IAK6kpHCMLMlwUMAwMVBIcLHx8slKyd4QqNio5HAARqHhltFxCUkpzGkZWTkN+SyFXjGVpAhW2KGJns2tmdkwABYI-AB0wSFdzD3MYH0DVuQjTePpk3Azc-PVy2NwSgpwyqrQ8JraugDK6eav8DbAdrqWxiaOMhUGiudx7OAbfqDRCoTg1OBmD71eKjZotA7tXKkFaXRS3cD3I7SMB4SjPcyWawgWz2fSGf6A5wguCRRJsADWwGkPEEwiQ4h4Umkl1ccAAvAjKV8fnSjFBTIzgW44AB+OAfCmGUV8CLYVhieRAA)

### 💻 Code

```typescript
export interface IFieldDef
{
    [name: string]: any;
}

export class Record<TRec extends IFieldDef>
{
    private a():TRec|null {
        return null;
    }
    protected b():TRec|null {
        return null;
    }
    public c() {
        return this.a();
    }
    public d() {
        return this.b();
    }
} 

export class Set<TSet extends Record<IFieldDef>>
{
    public item(): TSet|null {
        return null;
    }
}

export type DS<TRec extends Record<IFieldDef> | { [key:string]:any }> = TRec extends Record<IFieldDef> ? Set<TRec> : TRec[];
```

### 🙁 Actual behavior

Error: Type 'Record<IFieldDef> & TRec' does not satisfy the constraint 'Record<IFieldDef>'. Type 'Record<IFieldDef> & TRec' is missing the following properties from type 'Record<IFieldDef>': a, b

### 🙂 Expected behavior

No error

**Comment by RyanCavanaugh** [comment]
Reduced
```ts
declare class MyRecord {
    private a();
    b(): unknown;
}

declare class MySet<TSet extends MyRecord> {
    public item(): TSet;
}

type DS<TRec extends MyRecord | { [key: string]: unknown }> = TRec extends MyRecord ? MySet<TRec> : TRec[];
```
